### PR TITLE
SRCNN s4744784

### PR DESCRIPTION
**Pull request by harrisontolley**
Name: Harrison Tolley
Student ID: 47447842

Description:
In this project, we have implemented a super-resolution convolutional neural network (SRCNN) to upscale the resolution of images. This has been done with the MRI ADNI brain scan dataset.

File Structure:
- figures/ - Images produced from the project, including training images, loss graphs, PSNR results, and raw results.
- models/ - directory containing the train model paths that can be loaded.
- dataset.py - File contains functions for loading the training and testing dataset.
- train.py - File is responsible for training the model and outputting training loss/PSNR graphs. Also saves the state of the model after it's trained
- modules.py - File containing the model used in the training loop.
- predict.py - File which loads the saved model state, and upscales validation images, and saves a raw results figure.
- utils.py - This file contains constants, hyper paramaters, which are used in all other files. Also contains a function for computing PSNR loss. 
- README.md - This file contains the documentation of the project, containing details about the the model, training, and it's results.

To run this project:
If model files are not present, run train.py and wait for training to finish.
If model files are present, run predict.py to save raw results image.

### CODE REVIEW

**Comment by uqmarlonbran** [comment]
**This is an initial inspection - no action is required yet**

**Difficulty:** Normal

**General Comments:** Overall good attempt. Images look clearer than low-rez and in-line with other attempts. README quite descriptive and well laid out. Lots of examples. Would be good to have a figure of the network architecture. No validation set used. Test loss is lower than training loss. PSNR is computed on the average MSE over a whole epoch, which I don't think is quite right.

**[Recognition Problem]**
"modules.py":

- Nice implementation of PixelCNN. Also includes batch norm and dropout. Good use of Sequential block.

"dataset.py":

- Nice use of ImageFolder to deal with AD and NC folder structure.
- Provides some visualization for checking.

"train.py":

- Nice training loop, keeps track of train loss and validation loss. "Validation" set is simply the test set, no actual validation set used.
- Plots training/testing MSE and PSNR. Not sure if PSNR computed properly as it is calculated based on the average MSE at the end of each epoch, rather than image to image.
- Saves the model in checkpoints rather than checking for lowest validation loss.

"predict.py":

- Provides visualization of test images.
- No metrics on test dataset, but these are computed in the train script.

"README.MD":

- Good description of overall approach.
- No graphic for network architecture, but code demonstrated and described in text.
- Dependencies provided.
- Training and testing losses provided. 
- Can't figure out a reason for better validation scores than training? Code looks ok.
- Student acknowledges the use of Copilot and Chat-GPT.

**[Commit Log]**
Commit log looks ok.

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks missing in some files -1
## Recognition Problem 
Solves problem 
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage
Module present
Commenting
No Data leakage, but no validation set -1
Difficulty: Normal -5
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe good
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good

**Comment by shakes76** [comment]
To merge, need repo README restored. Does not affect grade.

**Comment by wangzhaomxy** [comment]
No feedback attempt and no feedback marks lost.
